In [1]:
import os
import sys
import pandas as pd
import numpy as np
import requests
import json
from pprint import pprint

In [3]:
os.getcwd()

'/Users/fbahena/Desktop/paco_growth/nlp_course/ia_hate/preprocessing/etl_miso'

In [6]:
os.listdir("../../data")

['CorpushaterNet.zip',
 '.DS_Store',
 'CorpusRNN.zip',
 'misocorpus.zip',
 'miso-seal.txt',
 'README.md',
 'raw_corpus_2M.zip',
 '__MACOSX',
 'miso_retrieval',
 'miso-ddss.txt',
 'misocorpus.txt',
 'miso-varw.txt']

In [2]:
data_dir = "../../data"
misodf = pd.read_csv(os.path.join(data_dir,"misocorpus.txt"))
misodf.head()

Twitter ID
0  1207093177225793536
1  1206528092837163009
2  1207765486433439748
3  1205401211429687296
4  1205400804703780865

In [3]:
misodf.shape

(6680, 1)

In [4]:
# split dataset into 5 chunks 

chunks = [np.array_split(chunk.iloc[:], 14) for chunk in np.array_split(misodf.iloc[:,0], 5)]

/Users/fbahena/.pyenv/versions/3.9.16/envs/nlpenv/lib/python3.9/site-packages/numpy/core/fromnumeric.py:59: FutureWarning: 'Series.swapaxes' is deprecated and will be removed in a future version. Please use 'Series.transpose' instead.
  return bound(*args, **kwds)


In [5]:
len(chunks[0])

14

In [6]:
# set twitter token
with open(os.path.join(data_dir,"twitter_api_bearer_token.txt"), "r") as f:
    bearer_token = f.readline()[:-1]


In [7]:
def bearer_oauth(r):
    """
    Method required by bearer token authentication.
    """

    r.headers["Authorization"] = f"Bearer {bearer_token}"
    r.headers["User-Agent"] = "v2TweetLookupPython"
    return r

def create_url(ids):
    #convert to string and then join with comma.
    formatted_ids = "ids="+",".join(list(map(lambda x: str(x),ids)))
    # You can adjust ids to include a single Tweets.
    # Or you can add to up to 100 comma-separated IDs
    url = "https://api.twitter.com/2/tweets?{}".format(formatted_ids)
    return url

def get_data(url):
    response = requests.request("GET", url, auth=bearer_oauth)
    print(response.status_code)
    if response.status_code != 200:
        raise Exception(
            "Request returned an error: {} {}".format(
                response.status_code, response.text
            )
        )
    return response.json()

In [9]:
url = create_url(chunks[0][:20])


In [114]:
# res= get_data(url)
resu = [{"id":x["id"],"text":x["text"]} for x in res["data"]]
pd.DataFrame(resu)

id                                               text
0   1206528092837163009  Voy hacer tutoriales para enseñarle a las que ...
1   1207765486433439748   @LaFallaras Usted es española, pero que payasa!!
2   1205400804703780865  @rouco64 @Delamasa123 Estos cobardes jueces pr...
3   1206657958014308354  @LaFallaras Quien te ha dicho a ti que no pued...
4   1205810735231700992  @GretaThunberg Que te calles ya niñata y aprov...
5   1206646560618094594                   @LaFallaras Payasa háztelo mirar
6   1205612670868770821            @LaFallaras Lo dejamos en perra en celo
7   1206306793699323906  Es vergonzosa la sentencia a los jugadores del...
8   1209067992845312000  @reyreysincorona Así son estas feminazis. Quie...
9   1209934659611746310  @LaFallaras Lávate primero cacho cerda... y qu...
10  1202074832726351873  No entiendo el papel de los gays defendiendo e...

In [13]:
chunks[0][1]

96     1207130751667494912
97     1207130248300711936
98     1208808202948808704
99     1204255007060705282
100    1205150270931509248
              ...         
187    1204566184596844544
188    1204484507505766401
189    1204473552847753217
190    1204231317132591105
191    1204386729295327232
Name: Twitter ID, Length: 96, dtype: int64

### 1st section

In [10]:
len(chunks[0])
section_data = []

In [128]:

for idx,chunk in enumerate(chunks[0]):
    print(f"processing subchunk {idx}")
    url = create_url(chunk.tolist())
    res= get_data(url)
    resu = [{"id":x["id"],"text":x["text"]} for x in res["data"]]
    section_data.append(resu)

processing subchunk 0
200
processing subchunk 1
200
processing subchunk 2
200
processing subchunk 3
200
processing subchunk 4
200
processing subchunk 5
200
processing subchunk 6
200
processing subchunk 7
200
processing subchunk 8
200
processing subchunk 9
200
processing subchunk 10
200
processing subchunk 11
200
processing subchunk 12
200
processing subchunk 13
200


In [132]:
len(section_data)

14

In [136]:
flat = [item for row in section_data for item in row]
df_section1 = pd.DataFrame(flat)

In [137]:
# df_section1.to_csv("misocorpus_1.csv")

Section 2 

In [141]:
len(chunks[1])
section_data = []

In [142]:
for idx,chunk in enumerate(chunks[1]):
    print(f"processing subchunk {idx}")
    url = create_url(chunk.tolist())
    res= get_data(url)
    resu = [{"id":x["id"],"text":x["text"]} for x in res["data"]]
    section_data.append(resu)

processing subchunk 0
200
processing subchunk 1
200
processing subchunk 2
200
processing subchunk 3
200
processing subchunk 4
200
processing subchunk 5
200
processing subchunk 6
200
processing subchunk 7
200
processing subchunk 8
200
processing subchunk 9
200
processing subchunk 10
200
processing subchunk 11
200
processing subchunk 12
200
processing subchunk 13
200


In [143]:
flat = [item for row in section_data for item in row]
df_section2 = pd.DataFrame(flat)

In [144]:
df_section2.shape

(867, 2)

In [ ]:
df_section2.to_csv("misocorpus_2.csv")

#### 3 section

In [15]:
len(chunks[2])
section_data = []

In [16]:
for idx,chunk in enumerate(chunks[2]):
    print(f"processing subchunk {idx}")
    url = create_url(chunk.tolist())
    res= get_data(url)
    resu = [{"id":x["id"],"text":x["text"]} for x in res["data"]]
    section_data.append(resu)

processing subchunk 0
200
processing subchunk 1
200
processing subchunk 2
200
processing subchunk 3
200
processing subchunk 4
200
processing subchunk 5
200
processing subchunk 6
200
processing subchunk 7
200
processing subchunk 8
200
processing subchunk 9
200
processing subchunk 10
200
processing subchunk 11
200
processing subchunk 12
200
processing subchunk 13
200


In [17]:
flat = [item for row in section_data for item in row]
df_section = pd.DataFrame(flat)

In [19]:
df_section.to_csv("misocorpus_3.csv")

section 4

In [30]:
len(chunks[3])
section_data = []

In [31]:
for idx,chunk in enumerate(chunks[3]):
    print(f"processing subchunk {idx}")
    url = create_url(chunk.tolist())
    res= get_data(url)
    resu = [{"id":x["id"],"text":x["text"]} for x in res["data"]]
    section_data.append(resu)

processing subchunk 0
200
processing subchunk 1
200
processing subchunk 2
200
processing subchunk 3
200
processing subchunk 4
200
processing subchunk 5
200
processing subchunk 6
200
processing subchunk 7
200
processing subchunk 8
200
processing subchunk 9
200
processing subchunk 10
200
processing subchunk 11
200
processing subchunk 12
200
processing subchunk 13
200


In [32]:
flat = [item for row in section_data for item in row]
df_section = pd.DataFrame(flat)

In [33]:
df_section.to_csv("misocorpus_4.csv")

In [37]:
df_section.head()

id                                               text
0  1208567700525584386  @edufeiok Bien que te gusta decir clientA y si...
1  1208567508095131648  @LeonorBohme Aquí queda más que claro que el f...
2  1208613007770697729  Te conocí este año, me gustabas mucho, pero er...
3  1208627871406862336  @sofiabenuzio @l123737 @vixiglesias Es una ram...
4  1208296054514487297  @Dra_Tanuki @guillemclua Además no era una dem...

#### section 5

In [39]:
len(chunks[4])
section_data = []

In [40]:
for idx,chunk in enumerate(chunks[4]):
    print(f"processing subchunk {idx}")
    url = create_url(chunk.tolist())
    res= get_data(url)
    resu = [{"id":x["id"],"text":x["text"]} for x in res["data"]]
    section_data.append(resu)

processing subchunk 0
200
processing subchunk 1
200
processing subchunk 2
200
processing subchunk 3
200
processing subchunk 4
200
processing subchunk 5
200
processing subchunk 6
200
processing subchunk 7
200
processing subchunk 8
200
processing subchunk 9
200
processing subchunk 10
200
processing subchunk 11
200
processing subchunk 12
200
processing subchunk 13
200


In [41]:
flat = [item for row in section_data for item in row]
df_section = pd.DataFrame(flat)

In [42]:
df_section.to_csv("misocorpus_5.csv")

In [51]:

files = [x for x in os.listdir(".") if x.startswith("miso")]


In [69]:
files

['misocorpus_1.csv',
 'misocorpus_3.csv',
 'misocorpus_2.csv',
 'misocorpus_5.csv',
 'misocorpus_4.csv']

In [56]:
df = pd.concat([pd.read_csv(file).iloc[:,1:] for file in files])

In [59]:
df.to_csv("misocorpus_full.csv",index=False)

In [63]:
df.shape

(4684, 2)

In [72]:
rows = 0
for file in files:
    print(file)
    d = pd.read_csv(file)
    print(d.shape)
    rows += d.shape[0]
rows

misocorpus_1.csv
(898, 3)
misocorpus_3.csv
(913, 3)
misocorpus_2.csv
(867, 3)
misocorpus_5.csv
(970, 3)
misocorpus_4.csv
(1036, 3)


4684

In [74]:
df.shape

(4684, 2)

Note that 4684 is less than the ids provided, it means tweets were deleted

Now lets verify the 3 remaining databases,

They should be contained in the full misocorpus ids one.

In [76]:
data_dir = "../../data"
misodf = pd.read_csv(os.path.join(data_dir,"misocorpus.txt"))
misodf.head()

Twitter ID
0  1207093177225793536
1  1206528092837163009
2  1207765486433439748
3  1205401211429687296
4  1205400804703780865

In [80]:
miso_dss = pd.read_csv(os.path.join(data_dir,"miso-ddss.txt"))
miso_dss.shape

(2004, 1)

In [89]:
lj_dss = miso_dss.merge(misodf, on='Twitter ID', how='left',indicator=True)
lj_dss[lj_dss["_merge"] == "left_only"].count()

Twitter ID    29
_merge        29
dtype: int64

In [90]:
miso_seal = pd.read_csv(os.path.join(data_dir,"miso-seal.txt"))
miso_seal.shape

(2764, 1)

In [91]:
lj_seal = miso_seal.merge(misodf, on='Twitter ID', how='left',indicator=True)
lj_seal[lj_seal["_merge"] == "left_only"].count()

Twitter ID    0
_merge        0
dtype: int64

In [92]:
miso_varw = pd.read_csv(os.path.join(data_dir,"miso-varw.txt"))
miso_varw.shape

(3572, 1)

In [93]:
lj_varw = miso_varw.merge(misodf, on='Twitter ID', how='left',indicator=True)
lj_varw[lj_varw["_merge"] == "left_only"].count()

Twitter ID    0
_merge        0
dtype: int64